File: backend/domain_monitor/topics.py
from __future__ import annotations

import re
from collections import Counter
from dataclasses import dataclass, field
from typing import Iterable, Optional

import config_loader

# Rule match strength bands. Deliberately not called "confidence" — there is no
# model here, only counted keyword hits.
STRENGTH_HIGH = "High"
STRENGTH_MEDIUM = "Medium"
STRENGTH_LOW = "Low"
STRENGTH_NONE = "None"

RELEVANCE_HIGH = "High"
RELEVANCE_MEDIUM = "Medium"
RELEVANCE_LOW = "Low"
RELEVANCE_NONE = "None"

_WORD_SPLIT = re.compile(r"[^a-z0-9]+")
_pattern_cache: dict[str, re.Pattern[str]] = {}


def _pattern(keyword: str) -> re.Pattern[str]:
    """Word-boundary matcher for one keyword or phrase."""
    cached = _pattern_cache.get(keyword)
    if cached is None:
        cached = re.compile(rf"(?<![a-z0-9]){re.escape(keyword.lower())}(?![a-z0-9])")
        _pattern_cache[keyword] = cached
    return cached


def normalize_text(value: Optional[str]) -> str:
    """Lowercase and flatten separators so URLs and titles match the same way."""
    if not value:
        return ""
    return " ".join(_WORD_SPLIT.split(value.lower())).strip()


def count_hits(text: Optional[str], keywords: Iterable[str]) -> int:
    """Total keyword occurrences in one text field."""
    haystack = normalize_text(text)
    if not haystack:
        return 0
    return sum(len(_pattern(kw).findall(haystack)) for kw in keywords)


@dataclass
class TopicSignal:
    """Which text fields produced hits for one topic, and how many."""

    topic: str
    title_hits: int = 0
    meta_hits: int = 0
    anchor_hits: int = 0
    url_hits: int = 0

    @property
    def total(self) -> int:
        return self.title_hits + self.meta_hits + self.anchor_hits + self.url_hits


@dataclass
class TopicResult:
    primary_topic: Optional[str] = None
    secondary_topics: list[str] = field(default_factory=list)
    topic_match_count: int = 0
    match_strength: str = STRENGTH_NONE
    per_topic: dict[str, int] = field(default_factory=dict)
    signals: dict[str, TopicSignal] = field(default_factory=dict)

    @property
    def has_data(self) -> bool:
        return bool(self.per_topic)


@dataclass
class TextEvidence:
    """Everything the topic and spam rules are allowed to read."""

    titles: list[str] = field(default_factory=list)
    metas: list[str] = field(default_factory=list)
    urls: list[str] = field(default_factory=list)
    anchors: list[str] = field(default_factory=list)

    @property
    def is_empty(self) -> bool:
        return not (self.titles or self.metas or self.urls or self.anchors)

    def joined(self) -> str:
        return " ".join([*self.titles, *self.metas, *self.urls, *self.anchors])


def _topic_vocabulary(include_spam: bool) -> dict[str, list[str]]:
    """Editorial topics, optionally widened with the spam categories.

    Spam categories count as topics for timeline and switch detection: a site
    that turned into a casino really did change topic, and treating that as
    "no topic" would hide the volatility the spam rules need to see.
    """
    vocabulary = dict(config_loader.topics())
    if include_spam:
        for name, spec in config_loader.spam_categories().items():
            vocabulary.setdefault(name, spec.get("keywords", []))
    return vocabulary


def classify(evidence: TextEvidence, include_spam: bool = True) -> TopicResult:
    """Score every configured topic against the collected text evidence.

    Returns None-valued topics when there is nothing to read, so the caller can
    distinguish "no topic matched" from "no data collected".
    """
    result = TopicResult()
    if evidence.is_empty:
        return result

    configured = _topic_vocabulary(include_spam)
    signals: dict[str, TopicSignal] = {}

    for topic, keywords in configured.items():
        signal = TopicSignal(topic=topic)
        for title in evidence.titles:
            signal.title_hits += count_hits(title, keywords)
        for meta in evidence.metas:
            signal.meta_hits += count_hits(meta, keywords)
        for anchor in evidence.anchors:
            signal.anchor_hits += count_hits(anchor, keywords)
        for url in evidence.urls:
            signal.url_hits += count_hits(url, keywords)
        if signal.total > 0:
            signals[topic] = signal

    if not signals:
        return result

    ranked = sorted(signals.values(), key=lambda s: (-s.total, s.topic))
    result.signals = signals
    result.per_topic = {s.topic: s.total for s in ranked}
    result.primary_topic = ranked[0].topic
    result.secondary_topics = [s.topic for s in ranked[1:4]]
    result.topic_match_count = ranked[0].total

    top = ranked[0].total
    fields_hit = sum(
        1 for value in (ranked[0].title_hits, ranked[0].meta_hits, ranked[0].anchor_hits, ranked[0].url_hits) if value
    )
    # Strength rises with both volume and how many independent fields agree.
    if top >= 8 and fields_hit >= 2:
        result.match_strength = STRENGTH_HIGH
    elif top >= 4:
        result.match_strength = STRENGTH_MEDIUM
    else:
        result.match_strength = STRENGTH_LOW

    return result


def relevance(
    evidence: TextEvidence,
    target_niches: Iterable[str],
) -> tuple[Optional[int], str]:
    """Weighted keyword score for the admin's target niches.

    Returns (score, band). Score is None when no niches are configured, which
    the UI renders as "not set" rather than as a zero.
    """
    niches = [n for n in target_niches if n]
    if not niches:
        return None, RELEVANCE_NONE
    if evidence.is_empty:
        return None, RELEVANCE_NONE

    cfg = config_loader.scoring().get("topical_relevance", {})
    points = cfg.get("points", {})
    cap = int(cfg.get("score_cap", 40))
    bands = cfg.get("bands", {})
    configured = config_loader.topics()

    score = 0
    for niche in niches:
        keywords = configured.get(niche)
        if not keywords:
            continue
        for title in evidence.titles:
            score += count_hits(title, keywords) * int(points.get("title_match", 5))
        for meta in evidence.metas:
            score += count_hits(meta, keywords) * int(points.get("meta_match", 3))
        for anchor in evidence.anchors:
            score += count_hits(anchor, keywords) * int(points.get("anchor_match", 3))
        for url in evidence.urls:
            score += count_hits(url, keywords) * int(points.get("url_match", 2))

    score = min(score, cap)

    if score >= int(bands.get("high_min", 20)):
        return score, RELEVANCE_HIGH
    if score >= int(bands.get("medium_min", 10)):
        return score, RELEVANCE_MEDIUM
    if score >= int(bands.get("low_min", 1)):
        return score, RELEVANCE_LOW
    return score, RELEVANCE_NONE


def topic_timeline(snapshots: list[dict]) -> list[dict]:
    """Per-snapshot primary topic, oldest first. Drives the history timeline."""
    timeline: list[dict] = []
    for snap in sorted(snapshots, key=lambda s: s.get("year") or 0):
        evidence = TextEvidence(
            titles=[snap.get("title") or ""],
            metas=[snap.get("meta_description") or ""],
            urls=[snap.get("url") or ""],
        )
        result = classify(evidence, include_spam=True)
        timeline.append(
            {
                "year": snap.get("year"),
                "timestamp": snap.get("timestamp"),
                "title": snap.get("title"),
                "topic": result.primary_topic,
            }
        )
    return timeline


def count_topic_switches(timeline: list[dict]) -> int:
    """Distinct consecutive topic changes across the timeline.

    Snapshots with no detectable topic are skipped rather than counted as a
    change, so a blank archive page does not manufacture volatility.
    """
    seen = [entry["topic"] for entry in timeline if entry.get("topic")]
    if len(seen) < 2:
        return 0
    return sum(1 for a, b in zip(seen, seen[1:]) if a != b)


def stability_label(switch_count: int) -> str:
    labels = config_loader.scoring().get("historical_stability", {}).get("stability_labels", {})
    if switch_count <= int(labels.get("stable_max_switches", 1)):
        return "Stable"
    if switch_count <= int(labels.get("some_changes_max_switches", 3)):
        return "Some Changes"
    return "High Topic Volatility"


def dominant_topic(timeline: list[dict]) -> Optional[str]:
    """Most frequent topic across the archive, used as the historical topic."""
    counts = Counter(entry["topic"] for entry in timeline if entry.get("topic"))
    if not counts:
        return None
    return counts.most_common(1)[0][0]

File: node-backend/domainMonitor/topics.js
"use strict";

// Faithful Node port of backend/domain_monitor/topics.py
const configLoader = require("./configLoader");

// Rule match strength bands. Deliberately not called "confidence" — there is no
// model here, only counted keyword hits.
const STRENGTH_HIGH = "High";
const STRENGTH_MEDIUM = "Medium";
const STRENGTH_LOW = "Low";
const STRENGTH_NONE = "None";

const RELEVANCE_HIGH = "High";
const RELEVANCE_MEDIUM = "Medium";
const RELEVANCE_LOW = "Low";
const RELEVANCE_NONE = "None";

const _WORD_SPLIT = /[^a-z0-9]+/;
const _patternCache = new Map();

function _escapeRegex(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/** Word-boundary matcher for one keyword or phrase. */
function _pattern(keyword) {
  let cached = _patternCache.get(keyword);
  if (cached === undefined) {
    cached = new RegExp(
      `(?<![a-z0-9])${_escapeRegex(keyword.toLowerCase())}(?![a-z0-9])`,
      "g"
    );
    _patternCache.set(keyword, cached);
  }
  return cached;
}

/** Lowercase and flatten separators so URLs and titles match the same way. */
function normalizeText(value) {
  if (!value) return "";
  return String(value).toLowerCase().split(_WORD_SPLIT).join(" ").trim();
}

/** Total keyword occurrences in one text field. */
function countHits(text, keywords) {
  const haystack = normalizeText(text);
  if (!haystack) return 0;
  let total = 0;
  for (const kw of keywords) {
    const re = _pattern(kw);
    re.lastIndex = 0;
    const matches = haystack.match(re);
    if (matches) total += matches.length;
  }
  return total;
}

/** Which text fields produced hits for one topic, and how many. */
class TopicSignal {
  constructor(topic) {
    this.topic = topic;
    this.title_hits = 0;
    this.meta_hits = 0;
    this.anchor_hits = 0;
    this.url_hits = 0;
  }

  get total() {
    return this.title_hits + this.meta_hits + this.anchor_hits + this.url_hits;
  }
}

/** Everything the topic and spam rules are allowed to read. */
class TextEvidence {
  constructor({ titles = [], metas = [], urls = [], anchors = [] } = {}) {
    this.titles = titles;
    this.metas = metas;
    this.urls = urls;
    this.anchors = anchors;
  }

  get is_empty() {
    return !(
      this.titles.length ||
      this.metas.length ||
      this.urls.length ||
      this.anchors.length
    );
  }

  joined() {
    return [...this.titles, ...this.metas, ...this.urls, ...this.anchors].join(
      " "
    );
  }
}

/**
 * Editorial topics, optionally widened with the spam categories.
 *
 * Spam categories count as topics for timeline and switch detection: a site
 * that turned into a casino really did change topic.
 */
function _topicVocabulary(includeSpam) {
  const vocabulary = { ...configLoader.topics() };
  if (includeSpam) {
    const cats = configLoader.spamCategories();
    for (const [name, spec] of Object.entries(cats)) {
      if (!Object.prototype.hasOwnProperty.call(vocabulary, name)) {
        vocabulary[name] = (spec && spec.keywords) || [];
      }
    }
  }
  return vocabulary;
}

/** Score every configured topic against the collected text evidence. */
function classify(evidence, includeSpam = true) {
  const result = {
    primary_topic: null,
    secondary_topics: [],
    topic_match_count: 0,
    match_strength: STRENGTH_NONE,
    per_topic: {},
    signals: {},
    has_data: false,
  };
  if (evidence.is_empty) return result;

  const configured = _topicVocabulary(includeSpam);
  const signals = {};

  for (const [topic, keywords] of Object.entries(configured)) {
    const signal = new TopicSignal(topic);
    for (const title of evidence.titles) signal.title_hits += countHits(title, keywords);
    for (const meta of evidence.metas) signal.meta_hits += countHits(meta, keywords);
    for (const anchor of evidence.anchors) signal.anchor_hits += countHits(anchor, keywords);
    for (const url of evidence.urls) signal.url_hits += countHits(url, keywords);
    if (signal.total > 0) signals[topic] = signal;
  }

  const signalList = Object.values(signals);
  if (signalList.length === 0) return result;

  const ranked = signalList.sort((a, b) => {
    if (b.total !== a.total) return b.total - a.total;
    return a.topic < b.topic ? -1 : a.topic > b.topic ? 1 : 0;
  });

  result.signals = signals;
  result.per_topic = {};
  for (const s of ranked) result.per_topic[s.topic] = s.total;
  result.primary_topic = ranked[0].topic;
  result.secondary_topics = ranked.slice(1, 4).map((s) => s.topic);
  result.topic_match_count = ranked[0].total;
  result.has_data = Object.keys(result.per_topic).length > 0;

  const top = ranked[0].total;
  const fieldsHit = [
    ranked[0].title_hits,
    ranked[0].meta_hits,
    ranked[0].anchor_hits,
    ranked[0].url_hits,
  ].filter((v) => v).length;

  // Strength rises with both volume and how many independent fields agree.
  if (top >= 8 && fieldsHit >= 2) result.match_strength = STRENGTH_HIGH;
  else if (top >= 4) result.match_strength = STRENGTH_MEDIUM;
  else result.match_strength = STRENGTH_LOW;

  return result;
}

/**
 * Weighted keyword score for the admin's target niches.
 *
 * Returns [score, band]. Score is null when no niches are configured.
 */
function relevance(evidence, targetNiches) {
  const niches = (targetNiches || []).filter((n) => n);
  if (niches.length === 0) return [null, RELEVANCE_NONE];
  if (evidence.is_empty) return [null, RELEVANCE_NONE];

  const cfg = configLoader.scoring().topical_relevance || {};
  const points = cfg.points || {};
  const cap = parseInt(cfg.score_cap ?? 40, 10);
  const bands = cfg.bands || {};
  const configured = configLoader.topics();

  let score = 0;
  for (const niche of niches) {
    const keywords = configured[niche];
    if (!keywords) continue;
    for (const title of evidence.titles)
      score += countHits(title, keywords) * parseInt(points.title_match ?? 5, 10);
    for (const meta of evidence.metas)
      score += countHits(meta, keywords) * parseInt(points.meta_match ?? 3, 10);
    for (const anchor of evidence.anchors)
      score += countHits(anchor, keywords) * parseInt(points.anchor_match ?? 3, 10);
    for (const url of evidence.urls)
      score += countHits(url, keywords) * parseInt(points.url_match ?? 2, 10);
  }

  score = Math.min(score, cap);

  if (score >= parseInt(bands.high_min ?? 20, 10)) return [score, RELEVANCE_HIGH];
  if (score >= parseInt(bands.medium_min ?? 10, 10)) return [score, RELEVANCE_MEDIUM];
  if (score >= parseInt(bands.low_min ?? 1, 10)) return [score, RELEVANCE_LOW];
  return [score, RELEVANCE_NONE];
}

/** Per-snapshot primary topic, oldest first. Drives the history timeline. */
function topicTimeline(snapshots) {
  const timeline = [];
  const sorted = [...snapshots].sort(
    (a, b) => (a.year || 0) - (b.year || 0)
  );
  for (const snap of sorted) {
    const evidence = new TextEvidence({
      titles: [snap.title || ""],
      metas: [snap.meta_description || ""],
      urls: [snap.url || ""],
    });
    const result = classify(evidence, true);
    timeline.push({
      year: snap.year ?? null,
      timestamp: snap.timestamp ?? null,
      title: snap.title ?? null,
      topic: result.primary_topic,
    });
  }
  return timeline;
}

/**
 * Distinct consecutive topic changes across the timeline.
 *
 * Snapshots with no detectable topic are skipped rather than counted.
 */
function countTopicSwitches(timeline) {
  const seen = timeline.filter((e) => e.topic).map((e) => e.topic);
  if (seen.length < 2) return 0;
  let switches = 0;
  for (let i = 1; i < seen.length; i++) {
    if (seen[i] !== seen[i - 1]) switches += 1;
  }
  return switches;
}

function stabilityLabel(switchCount) {
  const labels =
    (configLoader.scoring().historical_stability || {}).stability_labels || {};
  if (switchCount <= parseInt(labels.stable_max_switches ?? 1, 10)) return "Stable";
  if (switchCount <= parseInt(labels.some_changes_max_switches ?? 3, 10))
    return "Some Changes";
  return "High Topic Volatility";
}

/** Most frequent topic across the archive, used as the historical topic. */
function dominantTopic(timeline) {
  const counts = new Map();
  for (const entry of timeline) {
    if (entry.topic) counts.set(entry.topic, (counts.get(entry.topic) || 0) + 1);
  }
  if (counts.size === 0) return null;
  let best = null;
  let bestCount = -Infinity;
  for (const [topic, count] of counts) {
    if (count > bestCount) {
      best = topic;
      bestCount = count;
    }
  }
  return best;
}

module.exports = {
  STRENGTH_HIGH,
  STRENGTH_MEDIUM,
  STRENGTH_LOW,
  STRENGTH_NONE,
  RELEVANCE_HIGH,
  RELEVANCE_MEDIUM,
  RELEVANCE_LOW,
  RELEVANCE_NONE,
  normalizeText,
  countHits,
  TopicSignal,
  TextEvidence,
  classify,
  relevance,
  topicTimeline,
  countTopicSwitches,
  stabilityLabel,
  dominantTopic,
};
